# Лабораторная работа 6 Югай Ильи AIN-1-21

In [4]:
# Импортируем необходимые библиотеки
import pandas as pd
import numpy as np
import random
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.ensemble import RandomForestClassifier

In [5]:
# Задание 1: Изучить материал и примеры, описанные в ходе работы (датасет iris)
# Загружаем датасет iris
from sklearn.datasets import load_iris
iris = load_iris()
# Создаем датафрейм с признаками и целевой переменной
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)
iris_df['target'] = iris.target
# Разделяем датасет на обучающую и тестовую выборки
X_train_iris, X_test_iris, y_train_iris, y_test_iris = train_test_split(iris_df.drop('target', axis=1), iris_df['target'], test_size=0.2, random_state=42)
# Создаем модель классификатора на основе случайного леса
rf_iris = RandomForestClassifier(random_state=42)
# Обучаем модель на обучающей выборке
rf_iris.fit(X_train_iris, y_train_iris)
# Предсказываем классы для тестовой выборки
y_pred_iris = rf_iris.predict(X_test_iris)
# Оцениваем качество модели с помощью метрик
print('Accuracy score for iris dataset:', accuracy_score(y_test_iris, y_pred_iris))
print('Confusion matrix for iris dataset:\n', confusion_matrix(y_test_iris, y_pred_iris))
print('Classification report for iris dataset:\n', classification_report(y_test_iris, y_pred_iris))

Accuracy score for iris dataset: 1.0
Confusion matrix for iris dataset:
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]
Classification report for iris dataset:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30


In [6]:
# Задание 2: Используя один из подходящих игрушечных датасетов (breast_canser, digits, diabetes …) применить алгоритмы улучшения качества и оценки моделей (перекрестная проверка, решетчатый поиск, метрики модели)
# Загружаем датасет breast_cancer
breast_cancer = load_breast_cancer()
# Создаем датафрейм с признаками и целевой переменной
breast_cancer_df = pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)
breast_cancer_df['target'] = breast_cancer.target
# Разделяем датасет на обучающую и тестовую выборки
X_train_bc, X_test_bc, y_train_bc, y_test_bc = train_test_split(breast_cancer_df.drop('target', axis=1), breast_cancer_df['target'], test_size=0.2, random_state=42)
# Создаем модель классификатора на основе случайного леса
rf_bc = RandomForestClassifier(random_state=42)
# Обучаем модель на обучающей выборке
rf_bc.fit(X_train_bc, y_train_bc)
# Предсказываем классы для тестовой выборки
y_pred_bc = rf_bc.predict(X_test_bc)
# Оцениваем качество модели с помощью метрик
print('Accuracy score for breast cancer dataset:', accuracy_score(y_test_bc, y_pred_bc))
print('Confusion matrix for breast cancer dataset:\n', confusion_matrix(y_test_bc, y_pred_bc))
print('Classification report for breast cancer dataset:\n', classification_report(y_test_bc, y_pred_bc))
# Применяем перекрестную проверку для оценки модели на разных подвыборках
cv_scores = cross_val_score(rf_bc, breast_cancer_df.drop('target', axis=1), breast_cancer_df['target'], cv=5)
print('Cross-validation scores for breast cancer dataset:', cv_scores)
# Применяем решетчатый поиск для подбора оптимальных гиперпараметров модели
param_grid = {'n_estimators': [10, 50, 100, 200], 'max_depth': [None, 2, 4, 6, 8], 'min_samples_split': [2, 4, 6, 8]}
grid_search = GridSearchCV(rf_bc, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train_bc, y_train_bc)
print('Best parameters for breast cancer dataset:', grid_search.best_params_)
print('Best score for breast cancer dataset:', grid_search.best_score_)
# Создаем новую модель с оптимальными гиперпараметрами
rf_bc_opt = RandomForestClassifier(n_estimators=grid_search.best_params_['n_estimators'], max_depth=grid_search.best_params_['max_depth'], min_samples_split=grid_search.best_params_['min_samples_split'], random_state=42)
# Обучаем новую модель на обучающей выборке
rf_bc_opt.fit(X_train_bc, y_train_bc)
# Предсказываем классы для тестовой выборки
y_pred_bc_opt = rf_bc_opt.predict(X_test_bc)
# Оцениваем качество новой модели с помощью метрик
print('Accuracy score for breast cancer dataset (optimized):', accuracy_score(y_test_bc, y_pred_bc_opt))
print('Confusion matrix for breast cancer dataset (optimized):\n', confusion_matrix(y_test_bc, y_pred_bc_opt))
print('Classification report for breast cancer dataset (optimized):\n', classification_report(y_test_bc, y_pred_bc_opt))

Accuracy score for breast cancer dataset: 0.9649122807017544
Confusion matrix for breast cancer dataset:
 [[40  3]
 [ 1 70]]
Classification report for breast cancer dataset:
               precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114
Cross-validation scores for breast cancer dataset: [0.92105263 0.93859649 0.98245614 0.96491228 0.97345133]
Best parameters for breast cancer dataset: {'max_depth': None, 'min_samples_split': 2, 'n_estimators': 200}
Best score for breast cancer dataset: 0.9626373626373625
Accuracy score for breast cancer dataset (optimized): 0.9649122807017544
Confusion matrix for breast cancer dataset (optimized):
 [[40  3]
 [ 1 70]]
Classification report for breast cancer dataset (optimized):
               p

In [12]:
# Задание 3: Создать модель классификатор, которая по определенному обучаемому набору (salary, city, age, vacation_prefer, transport_prefer) будет определять предпочтения человека (target), в каком городе провести отпуск

# Задание 3a: Создать свой датасет с категориальными данными (1000 строк), которые содержат следующие поля: (salary, city, age, vacation_prefer, transport_prefer, target). Применить import random (random. choice, random.randint)
# Создаем списки возможных значений для каждого поля
salary_list = [50000, 60000, 70000, 80000, 90000, 100000]
city_list = ['Bishkek', 'Osh', 'Jalal-Abad', 'Karakol', 'Tokmok', 'Naryn']
age_list = list(range(30, 66))
vacation_prefer_list = ['Shopping', 'Beach holiday', 'Skiing', 'Hiking', 'Cultural tour', 'Relaxing']
transport_prefer_list = ['auto', 'plane', 'train', 'bus']
target_list = ['London', 'Moscow', 'Paris', 'New York', 'Dubai', 'Tokyo']
# Создаем пустой датафрейм с нужными колонками
vacation_df = pd.DataFrame(columns=['salary', 'city', 'age', 'vacation_prefer', 'transport_prefer', 'target'])
# Заполняем датафрейм случайными значениями из списков
for i in range(1000):
  vacation_df.loc[i] = [random.choice(salary_list), random.choice(city_list), random.choice(age_list), random.choice(vacation_prefer_list), random.choice(transport_prefer_list), random.choice(target_list)]
# Просматриваем первые пять строк датафрейма
vacation_df.head()

salary        city  age vacation_prefer transport_prefer    target
0   70000  Jalal-Abad   48          Hiking             auto     Tokyo
1   80000     Bishkek   54   Cultural tour              bus     Tokyo
2   60000     Bishkek   33          Skiing            train  New York
3   60000       Naryn   46   Cultural tour            train     Paris
4   60000       Naryn   60   Beach holiday             auto    Moscow

In [13]:
# Задание 3b: Salary – установить числовой тип (например 50000), city – город проживания (например, Bishkek), age – возраст (например, от 30 до 65), vacation_prefer – тип отдыха (например, Shopping или Beach holiday), transport_prefer – тип транспорта (например, auto, plane), target – город, в котором проведем отпуск (например, London, Moscow)
# Проверяем типы данных в датафрейме
vacation_df.dtypes
# Видим, что salary и age имеют тип int64, а остальные колонки имеют тип object
# Приводим salary к типу float64 для удобства
vacation_df['salary'] = vacation_df['salary'].astype(float)
# Просматриваем первые пять строк датафрейма
vacation_df.head()

salary        city  age vacation_prefer transport_prefer    target
0  70000.0  Jalal-Abad   48          Hiking             auto     Tokyo
1  80000.0     Bishkek   54   Cultural tour              bus     Tokyo
2  60000.0     Bishkek   33          Skiing            train  New York
3  60000.0       Naryn   46   Cultural tour            train     Paris
4  60000.0       Naryn   60   Beach holiday             auto    Moscow

In [14]:
# Продолжение кода лабораторной работы
# Задание 3c: Преобразовать категориальные данные в числовые, используя panadas.get_dummies
# Применяем функцию get_dummies к колонкам с категориальными данными
vacation_df = pd.get_dummies(vacation_df, columns=['city', 'vacation_prefer', 'transport_prefer'])
# Просматриваем первые пять строк датафрейма
vacation_df.head()

salary  age    target  city_Bishkek  city_Jalal-Abad  city_Karakol  \
0  70000.0   48     Tokyo         False             True         False   
1  80000.0   54     Tokyo          True            False         False   
2  60000.0   33  New York          True            False         False   
3  60000.0   46     Paris         False            False         False   
4  60000.0   60    Moscow         False            False         False   

   city_Naryn  city_Osh  city_Tokmok  vacation_prefer_Beach holiday  \
0       False     False        False                          False   
1       False     False        False                          False   
2       False     False        False                          False   
3        True     False        False                          False   
4        True     False        False                           True   

   vacation_prefer_Cultural tour  vacation_prefer_Hiking  \
0                          False                    True   
1                           True                   False   
2                          False                   False   
3                           True                   False   
4                          False                   False   

   vacation_prefer_Relaxing  vacation_prefer_Shopping  vacation_prefer_Skiing  \
0                     False                     False                   False   
1                     False                     False                   False   
2                     False                     False                    True   
3                     False                     False                   False   
4                     False                     False                   False   

   transport_prefer_auto  transport_prefer_bus  transport_prefer_plane  \
0                   True                 False                   False   
1                  False                  True                   False   
2                  False                 False                   False   
3                  False                 False                   False   
4                   True                 False                   False   

   transport_prefer_train  
0                   False  
1                   False  
2                    True  
3                    True  
4                   False

In [15]:
# Задание 3d: Выделить обучающую выборку и тестовую выборку (X_train, y_train, X-test, y_test)
# Разделяем датафрейм на признаки и целевую переменную
X = vacation_df.drop('target', axis=1)
y = vacation_df['target']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [16]:
# Задание 3e: Выбрать модель классификатор (например, from sklearn.ensemble import RandomForestClassifier)
# Создаем модель классификатора на основе случайного леса
rf_vacation = RandomForestClassifier(random_state=42)

In [19]:
# Задание 3f: Проверить оценку модели
# Обучаем модель на обучающей выборке
rf_vacation.fit(X_train, y_train)
# Предсказываем классы для тестовой выборки
y_pred_vacation = rf_vacation.predict(X_test)
# Оцениваем качество модели с помощью метрик
print('Accuracy score for vacation dataset:', accuracy_score(y_test, y_pred_vacation))
print('Confusion matrix for vacation dataset:\n', confusion_matrix(y_test, y_pred_vacation))
print('Classification report for vacation dataset:\n', classification_report(y_test, y_pred_vacation))

Accuracy score for vacation dataset: 0.165
Confusion matrix for vacation dataset:
 [[7 6 2 9 5 8]
 [7 1 5 5 4 3]
 [5 2 5 1 9 5]
 [7 6 7 6 8 6]
 [9 3 9 3 8 5]
 [8 5 3 7 5 6]]
Classification report for vacation dataset:
               precision    recall  f1-score   support

       Dubai       0.16      0.19      0.18        37
      London       0.04      0.04      0.04        25
      Moscow       0.16      0.19      0.17        27
    New York       0.19      0.15      0.17        40
       Paris       0.21      0.22      0.21        37
       Tokyo       0.18      0.18      0.18        34

    accuracy                           0.17       200
   macro avg       0.16      0.16      0.16       200
weighted avg       0.16      0.17      0.16       200


In [20]:
# Задание 3g: Сделать предсказание на случайных данных
# Создаем случайные данные с теми же колонками, что и в датафрейме
random_data = pd.DataFrame(columns=X.columns)
random_data.loc[0] = [random.choice(salary_list), random.choice(age_list)] + [random.randint(0, 1) for i in range(len(random_data.columns) - 2)]
# Просматриваем случайные данные
random_data
# Предсказываем класс для случайных данных
random_pred = rf_vacation.predict(random_data)
# Выводим предсказанный класс
print('Predicted class for random data:', random_pred)

Predicted class for random data: ['Paris']


In [21]:
# Задание 3h: Улучшить модель согласно примерам лекций 9-10 и материала лабораторной работы
# Применяем перекрестную проверку для оценки модели на разных подвыборках
cv_scores_vacation = cross_val_score(rf_vacation, X, y, cv=5)
print('Cross-validation scores for vacation dataset:', cv_scores_vacation)
# Применяем решетчатый поиск для подбора оптимальных гиперпараметров модели
param_grid_vacation = {'n_estimators': [10, 50, 100, 200], 'max_depth': [None, 2, 4, 6, 8], 'min_samples_split': [2, 4, 6, 8]}
grid_search_vacation = GridSearchCV(rf_vacation, param_grid_vacation, cv=5, scoring='accuracy')
grid_search_vacation.fit(X_train, y_train)
print('Best parameters for vacation dataset:', grid_search_vacation.best_params_)
print('Best score for vacation dataset:', grid_search_vacation.best_score_)

Cross-validation scores for vacation dataset: [0.16  0.145 0.16  0.165 0.17 ]
Best parameters for vacation dataset: {'max_depth': 2, 'min_samples_split': 2, 'n_estimators': 200}
Best score for vacation dataset: 0.1775


In [25]:
# Создаем новую модель с оптимальными гиперпараметрами
rf_vacation_opt = RandomForestClassifier(n_estimators=grid_search_vacation.best_params_['n_estimators'], max_depth=grid_search_vacation.best_params_['max_depth'], min_samples_split=grid_search_vacation.best_params_['min_samples_split'], random_state=42)
# Обучаем новую модель на обучающей выборке
rf_vacation_opt.fit(X_train, y_train)
# Предсказываем классы для тестовой выборки
y_pred_vacation_opt = rf_vacation_opt.predict(X_test)
# Оцениваем качество новой модели с помощью метрик
print('Accuracy score for vacation dataset (optimized):', accuracy_score(y_test, y_pred_vacation_opt))
print('Confusion matrix for vacation dataset (optimized):\n', confusion_matrix(y_test, y_pred_vacation_opt))
print('Classification report for vacation dataset (optimized):\n', classification_report(y_test, y_pred_vacation_opt, zero_division=1))

Accuracy score for vacation dataset (optimized): 0.17
Confusion matrix for vacation dataset (optimized):
 [[ 9  0 21  0  7  0]
 [ 4  0 16  0  5  0]
 [ 4  0 14  1  8  0]
 [ 9  0 16  0 15  0]
 [ 8  0 17  1 11  0]
 [ 6  0 20  1  7  0]]
Classification report for vacation dataset (optimized):
               precision    recall  f1-score   support

       Dubai       0.23      0.24      0.23        37
      London       1.00      0.00      0.00        25
      Moscow       0.13      0.52      0.21        27
    New York       0.00      0.00      1.00        40
       Paris       0.21      0.30      0.24        37
       Tokyo       1.00      0.00      0.00        34

    accuracy                           0.17       200
   macro avg       0.43      0.18      0.28       200
weighted avg       0.39      0.17      0.32       200


## Вывод из работы
#### В ходе лабораторной работы я изучил материал и примеры по оценке качества модели на датасете iris. Я применил алгоритмы улучшения качества и оценки моделей (перекрестная проверка, решетчатый поиск, метрики модели) на датасете breast_cancer. Я создал свою модель классификатора, которая по определенному обучаемому набору (salary, city, age, vacation_prefer, transport_prefer) определяет предпочтения человека (target), в каком городе провести отпуск. Я сгенерировал свой датасет с категориальными данными, преобразовал их в числовые, разделил на обучающую и тестовую выборки, обучил модель, проверил ее оценку, сделал предсказание на случайных данных и улучшил модель с помощью перекрестной проверки и решетчатого поиска. Я сравнил качество модели до (0.165) и после оптимизации (0.17) и увидел, что оно улучшилось.